037-kafka-canary-roadmap.md

```diff
@@ -0,0 +1,109 @@
+* The Strimzi community's lack of Golang expertise (or indeed enthusiasm) means it is somewhat unloved.
+
+## Proposal
+
+### Re-implement in Java
+
+The Kafka project itself only provides support for JVM clients.
+Naturally those clients are extremely well-tested, with both current and older broker versions. 
+So generally speaking, we would expect the Canary to experience fewer issues due to client bugs.
+It would also necessarily support the latest versions of the Kafka protocol, which would enable support for more features.
+
+The Canary also stands out as being the only Strimzi project that's written in Golang.
+There's nothing wrong with Golang, but there are non-technical drawbacks to this situation.
+Most existing Strimzi developers have more experience with Java than Golang.
+Some don't really know Golang at all.
+It's difficult to motivate people to use another language if they're using it only for one relatively small part of their job.
+Likewise, existing Golang developers are not going to be drawn to contributing to Strimzi on account of a single Golang component.
+So over all we expect our community would be better able to engage with the Canary component if it was implemented in Java.
+
+### Desirable features
+
+There are several client features we would like to be able use.
+
+* Support for SASL `OAUTHBEARER` mechanism. Sarama supports this.
+* Support for forcing leader elections. Sarama doesn't support this.
+* Support for transactional produce. Sarama doesn't seem to support this, at least not in a usable way. You can send the transaction-related requests, but there is no client library support akin to the `TransactionManager`, for ensuring the RPCs are sent correctly. 
+
+### Differential observations
+
+By running multiple producers and/or consumers, a canary could more accurately indicate where problems lay. 
+
+For example, in a cluster with multiple listeners using different authentication mechanisms, if produce requests via a `SASL/OAUTHBEARER`-authenticated listener fail, but succeed via the `SASL/PLAIN`-authenticated listener you can infer that the problem is with the login mechanism. 
+
+There are many dimensions in which you can do this: 
+
+* DNS resolution
+* TLS (e.g. different versions or detecting imminent broker certificate expiry)
+* Producing to replicas on different volumes
+* With/without transactional producers, to flush out problems with transaction coordinators.
+* Similarly with/without using consumer groups. Using different group ids to detect problems with group coordinators.
+
+
+### Actionable documentation
+
+Assuming the differential observations feature proves to provide worthwhile information, we would proceed to also provide documented procedures recommending corrective actions based on the specific faults it reported.
+
+### Changes and removals
+
+* Logging mechanism will be changed to Log4j2, to have it consistent with other Strimzi components
```

Comment:
"will use" ... it's clear it "will be changed" because we are moving from Golang to Java :-)